includes inactive locales to the TranslatorAwareType - it now allows to translate Shop parameters -> Product settings -> Products stock block with inactive languages as well

tomas862 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | On the currently released PrestaShop version when you disable the language <br> you can still configure it in back-office. Usually people make various modifications when the languages are disabled to them.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | 1. Install another language to have at least two languages enabled. <br> 2. Check the **International -> Localization -> Configuration block** - you will notice that both languages are displayed which is good. <br> 3. Disable the installed language and repeat step 2 - you will notice that the disabled language is missing <br> 4. If you check currently released version of PrestaShop, like 1.7.4.0 you will notice that even if the language is disabled you can apply any actions to it.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/9427)
<!-- Reviewable:end -->

sarjon committed:
CO: add new actionFrontControllerSetVariables hook

Commit:
Changes to improve the product controller and enable more than 62 combinations

Commit:
Edit the test and check PSR2

Commit:
BO: Several improvements and refactoring in the product Controller

Commit:
BO: Several improvements and refactoring in the product Controller

Commit:
BO: Several improvements and refactoring in the product Controller

Commit:
BO: Several improvements and refactoring in the product Controller

mickaelandrieu committed:
Extracted private functions into dedicated services

mickaelandrieu committed:
Extracted Product Categories form outside of Controller

mickaelandrieu committed:
Re-introduct ProductController:renderField action (BC)

mickaelandrieu committed:
Address comments on FilterCategoriesRequestPurifier class

mickaelandrieu committed:
Extracted export action into specific service

mickaelandrieu committed:
Used trans helper instead of calling translator

mickaelandrieu committed:
Fixed coding style issues

mickaelandrieu committed:
Introduced getAdminLink helper

mickaelandrieu committed:
Updated deprecation message

mickaelandrieu committed:
Reintroduced missing AdminModelAdapter:getAttributesResume function

mickaelandrieu committed:
CO: address @eternoendless comments

Commit:
start working on sql manager migration

Commit:
add request rendering in list

Commit:
add request sql list filtering

Commit:
add request sql settings form handler & form rendering

Commit:
implement settings save action

Commit:
update list content with alerts

Commit:
create request sql form

Commit:
implement table sorting & pagination

Commit:
add updated webpack config

Commit:
add filters reset

Commit:
implement request sql create action

Commit:
add request sql edit

Commit:
created request sql deletion from list

Commit:
add db tables & columns display in request sql form

Commit:
allow adding db table from tables select to sql

Commit:
allow adding table columns to sql query

Commit:
fix request sql settings save

Commit:
update translation domains

Commit:
implement request sql query validation

Commit:
update request sql form with cancel button

Commit:
added view row action of request sql

Commit:
add not found object handling

Commit:
add request sql export action

Commit:
add export action for request sql

mickaelandrieu committed:
Address @Quetzacoalt91 comment

mickaelandrieu committed:
Fixed undefined  value

mickaelandrieu committed:
Fixed display of combinations action

Commit:
sql manager improvements

Commit:
fix services & routing definition

Commit:
add security & demo annotations for controller

Commit:
fix sql manager services

Commit:
add request sql grid definition

Commit:
add row, bulk and grid actions for request sql grid

Commit:
add route params for SimpleRowActions

Commit:
update naming to RequestSql

Commit:
remove FormHandler interface from RequestSql entity form handling

Commit:
implement submitable bulk actions

Commit:
implement bulk action delete for request sql

Commit:
move CharsetEncoding class to Core

Commit:
update row action naming

Commit:
configure row action options in setOptions() method

Commit:
fix grid.js

Commit:
update requestsql grid dedfinition factory

Commit:
remove legacy controller

Commit:
small improvements

Commit:
small improvements

Commit:
docblock fixes

Commit:
clean up FrameworkBundleAdminController

Commit:
add search system

Commit:
update request grid definition with new column names

Commit:
add search reset url to request sql grid definition

Commit:
update request sql definition with redirect url

Commit:
fix flash messages on request sql deletion

Commit:
address @PierreRambaud commnets

Commit:
fox show sql query

Commit:
fix sample import file downloading

Commit:
fix available fields ajax load

PierreRambaud committed:
First refactoring, add docker-compose.yml and some refactoring

PierreRambaud committed:
Update errors with cli install

eternoendless committed:
Update style of forms on new pages

eternoendless committed:
Improve style of help boxes

eternoendless committed:
Improve label help box

Removed useless title (same as label) and added text for screen readers

eternoendless committed:
Use horizontal layout for preferences page

eternoendless committed:
Update UI kit to 1.0.2

PierreRambaud committed:
Run tests under docker

PierreRambaud committed:
Allow usage of mounted volume (or not)

PierreRambaud committed:
Allow to customize run-tests parameters

mickaelandrieu committed:
Improved according to @eternoendless comments

PierreRambaud committed:
Add DOCKER.md

PierreRambaud committed:
Little refactoring, indentation

eternoendless committed:
Fix wrong indentation

PierreRambaud committed:
Fix mysql, unable to connect

PierreRambaud committed:
Install composer dependencies

PierreRambaud committed:
Set timeout to 360 and fix debconf warnings

PierreRambaud committed:
Edit README

mickaelandrieu committed:
Renamed services according to conventions

mickaelandrieu committed:
Improved deprecation error message

sarjon committed:
rebase fixes

sarjon committed:
add new js grid and requred extensions

sarjon committed:
get export sql name from page breadcrumb

sarjon committed:
style row actions

sarjon committed:
fix codacy warnings

sarjon committed:
hide sensitive data in sql view

sarjon committed:
fix sql export in modal

sarjon committed:
build assets

sarjon committed:
fix SearchAndResetType naming

sarjon committed:
fix SQL ORDER keyword check if RequestSql model

sarjon committed:
handle PrestaShopDatabaseException when validating sql

eternoendless committed:
Build assets

jolelievre committed:
Add missing appendArray in OrderReturnLazyArray (BOOM-6039)

The FO page presenting the OrderReturn did not work anymore due to
modifications on the OrderReturnPresenter::present method which no
loner returns an array but a OrderReturnLazyArray object which is
perfect to access the generated fields like detail_url through the
getDetailsUrl method. But it removed to access to original array
data like state_name and such.

Rather than implementing all the getter methods we simply use the
AbstractLazyArray::appendArray method to open access to all the
missing array fields.

azisyus committed:
Update Address.php

PierreRambaud committed:
Update Dockerfile

Commit:
includes inactive locales also to the TranslatorAwareType

sarjon commented:
as far as i understand this change affects all child types? if so, this may affect a lot of forms in Back Office that uses it as a parent, am i right?

tomas862 commented:
@sarjon yes, exactly.

sarjon commented:
> @sarjon yes, exactly.

from what i see, this changes allows you to translate disabled languages. For example here: https://prnt.sc/kgxyak Not sure if thats an issue.

marionf commented:
Yes it's a problem because the translation of the disabled language is not saved

![screenrecord8](https://user-images.githubusercontent.com/13449658/43947348-1eaf00b4-9c88-11e8-8ab2-14479c4f1ba6.gif)

PierreRambaud committed:
Fix BOOM-6223 and little refactoring

PierreRambaud committed:
Use multiline for DIRECTORY_SEPARATOR

tony-novius committed:
[CO] Add hook to custom SwiftMessage before sending mail

jolelievre committed:
add formatted field file_size_formatted in attachments

tomas862 commented:
I believe the translations should be applied even if they are disabled. I'll re-check the problematic areas and commit fixes.

Commit:
adds new function in configuration which with additional parameter provided gets values from configuration with disabled language as well, uses it in Stock configuration page

Commit:
adds missing function

tomas862 commented:
Please check the usage of the new function in Product preferences page. The are plenty more controllers left to fix but I want to check if the usage of the new created function is good. 
The Configuration->get() adapter did not handled the languages which are in active. As in previous versions of PrestaShop everything was handled with HelperOptions by passing parameter 'textLang' for instance. 

Right now, in the same Configuration adapter I created new adapter for getInternational config and added extra argument which defines if we want to select enabled or disabled language.

@sarjon, @Quetzacoalt91  What do you think about this approach?

sarjon commented:
well, im not big fan of another public method on configuration service, but let's wait for other reviews.

p.s. if i understand correctly, you want to add disabled languages to this input https://prnt.sc/kh0y5p?

tomas862 commented:
Well it's basically an adapter for legacy functionality with an extra argument - since it's related with TranslatorAwareType maybe it's better to add a function there.

Yes and many more inputs. If you check how it was before then it's the part that is missing starting from PrestaShop version 1.7.4.0 - since people often first have language disabled then translate all the inputs and re-enable language I believe this functionality is required.

Yes, let's wait more reviews because I found more places there this functionality is needed.

sarjon committed:
init filters decouple from columns

sarjon committed:
update Customer Settings form layout

eternoendless committed:
Make it unnecessary to append a '_' to the controller name when using the AdminSecurity annotation

The voter now adds the underscore automatically if necessary.

eternoendless committed:
Remove unnecessary trailing underscore in controller names.

This is now handled automatically by PageVoter.

jolelievre committed:
add customizations field in product whitelist

sarjon committed:
add column association on filter

MathiasReker committed:
Remove unused taxes

sarjon committed:
fix Customer Settings form layout update

sarjon committed:
update Order Settings form layout

sarjon committed:
update Product Settings form layout

sarjon committed:
add blocks for product preference templates

sarjon committed:
update Maintenance form layout

sarjon committed:
update Performance form layout

sarjon committed:
in Performance form use class to toggle cache options visibility instead of modifying style on element

sarjon committed:
update Administration form layout

sarjon committed:
update E-mail form layout

sarjon committed:
update Localization form layout

sarjon committed:
update Geolocation form layout

sarjon committed:
update Shipping > Preferences form layout

sarjon committed:
update Invoices form layout

sarjon committed:
update Delivery Slips form layout

sarjon committed:
update grid filters redering

sarjon committed:
add missing methods for filter interfaces

sarjon committed:
remove setters from grid definition

sarjon committed:
make grid definition factory responsible for dispatching definition modification hook

tolispy committed:
Fix Error 500 from API request with PHP 7.2

line 800 assumes that self::$shopIDs is an array that can be counted.
But self::$shopIDs is null and the following error appears:
[PHP Warning #2] count(): Parameter must be an array or an object that implements Countable (/home/jojo/public_html/mydomain.com/presta2/classes/webservice/WebserviceRequest.php, line 800)

tolispy committed:
Fix 403 error when accessing localhost using IPv6 (Maxmind Database exists)

You can not access prestashop frontend, when visiting localhost, after install Maxmind Database.
This is happening only when IPv6 is activated, and localhost is resolving to IPv6.
Then displayRestrictedCountryPage() function is firing at line 472 in version 1.7.4.2.

This fix apply the same logic, that is already applied when we visit localhost using IPv4

Commit:
update sql manager options form layout

Commit:
update forms layout

Commit:
update responsiveness of form

PierreRambaud committed:
Change Shebang to use bash instead of sh

jolelievre committed:
change out of stock condition with greater or equals

jolelievre committed:
fix welcome page variable injection by adding missing %

mickaelandrieu commented on src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:
```diff
@@ -235,7 +235,7 @@ services:
         arguments:
             - "@translator"
-            - "@=service('prestashop.adapter.legacy.context').getLanguages()"
+            - "@=service('prestashop.adapter.legacy.context').getLanguages(false)"
```
good luck to understand what this false do here :'(

mickaelandrieu commented on src/Adapter/Configuration.php:
```diff
@@ -85,6 +86,31 @@ public function add(array $parameters = array())
+     *
+     * @return array - array key is the language id
+     */
+    public function getInternational($key, $activeLanguage = true, $idShopGroup = null, $idShop = null)
```
I disagree with this: how about a new class called `MultiLangConfiguration` with `get` and `getIncludingInactiveLocales()` functions?

```php
namespace PrestaShop\PrestaShop\Adapter\Configuration;

use Configuration as ConfigurationLegacy;
use Language as LanguageLegacy;

class MultiLangConfiguration implements MultiLangConfigurationInterface
{
    public function __contruct(ConfigurationLegacy $configuration)
    {
        $this->configuration = $configuration;
    }

    /* ... */ put your updated methods here.
} 
```

> As a general rule, we forbid boolean flags on new methods.

mickaelandrieu reviewed (changes requested):
minor changes to be adressed

mickaelandrieu committed:
added PHPDocs

mickaelandrieu committed:
fixed typo

PierreRambaud committed:
Remove pushd & popd and use sh for Ubuntu compatibilities

PierreRambaud committed:
Change wait-for-it script

sarjon committed:
remove filter definition from columns

sarjon committed:
update Logs grid definition

tomas862 committed:
uses legacy context to get both active and inactive language choices in translationAwareType

tomas862 committed:
removes legacy context dependency

sarjon committed:
update tests

tomas862 committed:
uses specific adapter for context language in translations aware type, changes subclasses inheritance

sarjon commented on src/Adapter/Language/ContextLanguageDataProvider.php:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @return array
+     */
+    public function getIncludingInactiveLocales()
```
how about `getInstalledLocales()`?

tomas862 replied:
possibly a better name, thanks :)

sarjon commented on src/Adapter/Language/ContextLanguageDataProvider.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Class ContextLanguageDataProvider is responsible for getting languages related data from legacy content class
+ */
+class ContextLanguageDataProvider
```
Theres already `PrestaShop\PrestaShop\Adapter\Language\LanguageDataProvider`, maybe you can update it?

Or you could change naming to `LocaleProvider`, what do you think?

tomas862 replied:
I thought at first but it's strictly related with the legacy context so moving to LanguageDataProvider might confuse someone - decided to name adapter so it would be clear that the language retrieved from it is related with the shops context.

sarjon commented on src/PrestaShopBundle/Form/Admin/Type/TranslatorAwareType.php:
```diff
@@ -38,26 +39,36 @@
+    ) {
         $this->translator = $translator;
-        $this->locales = $locales;
+        $this->locales = $languageDataProvider->getActiveLocales();
```
```diff
+ $this->activeLocales = $languageDataProvider->getActiveLocales();
- $this->locales = $languageDataProvider->getActiveLocales();
```

sarjon commented on src/PrestaShopBundle/Form/Admin/Type/TranslatorAwareType.php:
```diff
@@ -38,26 +39,36 @@
         $this->translator = $translator;
-        $this->locales = $locales;
+        $this->locales = $languageDataProvider->getActiveLocales();
+        $this->allLocales = $languageDataProvider->getIncludingInactiveLocales();
```
```diff
+ $this->installedLocales = $languageDataProvider->getInstalledLocales();
- $this->locales = $languageDataProvider->getIncludingInactiveLocales();
```

maybe is clearer to use `installedLocales` instead of `allLocales`?

sarjon commented on src/Adapter/LegacyContext.php:
```diff
@@ -206,6 +206,28 @@ public function getLanguages($active = true, $id_shop = false, $ids_only = false
+        $languages = Language::getLanguages(false);
+
+        $defaultLanguageFirst = $this->getLanguage();
+        usort($languages, function ($a, $b) use ($defaultLanguageFirst) {
```
why employee language has to be first? :thinking:

tomas862 replied:
was part of legacy stuff - did not wanted to change that :D

Quetzacoalt91 replied:
It allows all the BO forms to display the default language first.
Then, the additional selectors allow the merchant to set a string for another language.

tomas862 committed:
adds an adapter for multi-language configuration, uses it in product stock form on the product settings menu

mickaelandrieu commented on src/Adapter/Language/ContextLanguageDataProvider.php:
```diff
@@ -24,7 +24,7 @@ public function __construct(LegacyContext $legacyContext)
      * @return array
      */
-    public function getIncludingInactiveLocales()
+    public function getInstalledLocales()
```
you cant' change public API :-1:

tomas862 replied:
I'm not sure I fully understand the problem here. I created here new adapter class called 'ContextLanguageDataPeovider' which has two public methods: 
'getActiveLocales' and 'getInstalledLocales' - I changed the name from 'getIncludingInactiveLocales' to 'getInstalledLocales' to keep it shorter.

The name is incorrect here or something else is wrong? Let me know, thanks ;)

mickaelandrieu replied:
Is it a new class you created? forget my comment then :stuck_out_tongue:

PierreRambaud committed:
Change running command

mickaelandrieu commented:
> Yes, let's wait more reviews because I found more places there this functionality is needed.

Can you explain what is the functional issue to @colinegin? I'm not the right person to decide if a feature needs to be (re) introduced to the core.

mickaelandrieu committed:
reverted the strict condition

tomas862 commented:
@colinegin 

When merchant has multiple languages and decides to disable one language the translation dropdowns should be still visible in both languages in some places (Shop parameters -> Product settings -> Products stock block is currently included place but there are more places where such feature is in demand. Right now I implemented only here but later on I can include in more pages where it was used). This is I think required feature due to most merchants disable the language , apply all translations and re-enable it so all translations applied would be visible for regular customer on the shop in the front page. :)

mickaelandrieu committed:
Updated hook name

sarjon committed:
update search form handling

sarjon committed:
rename 'row' to 'record'

Commit:
fix templates

mickaelandrieu committed:
Address @eternoendless comment: remove extract() call

mickaelandrieu committed:
Removed all translations errors from localization page

mickaelandrieu committed:
Enable Logs page

mickaelandrieu committed:
Removed legacy controller, templates and test

Commit:
fix grid presenter test

mickaelandrieu committed:
Now rely on Grid component for the Search form

colinegin commented:
Hello @tomas862,

This is great news !
I think this is a good thing to put back all translation fields even for disabled languages.
Just be careful, if you put it back on Symfony pages where some fields are mandatory, we need to put back the same behavior as on legacy pages : if the merchant has 2 languages and fills only one translation, the same content is applied to the other language.

For optional fields, you can keep it the same way as you did.

Let me know if you have any questions !

Thank you very much :)

tomas862 committed:
removes dependency from an adapter class instance call

tomas862 committed:
changed child classes constructor parameters

tomas862 committed:
removed unused function

tomas862 committed:
changes iteration over active locales

tomas862 commented on src/PrestaShopBundle/Form/Admin/Type/TranslatorAwareType.php:
```diff
@@ -111,4 +116,23 @@ private function formatLocales(array $locales)
+        $result = [];
+        foreach ($locales as $locale) {
+            if ($locale['active']) {
+                $result['active'][] = $locale;
```
This one is fixed in the next commit https://github.com/PrestaShop/PrestaShop/pull/9427/commits/672d5938e7f5f416e701130f54023e9e7d922fcd

mickaelandrieu committed:
Rebased to develop branch

mickaelandrieu committed:
Fixed conflict from rebase

mickaelandrieu committed:
Fixed survival tests

eternoendless committed:
Improved readme

eternoendless committed:
Improve contributing file

sarjon committed:
fix request sql filters

mickaelandrieu committed:
Definition have no knowledge of view

eternoendless committed:
Delete unused property

eternoendless committed:
Add undeclared properties

eternoendless committed:
Fix incorrect paths in robots.txt

Until now, robots.txt was blocking not only directories at the root,
but also directories named the same in any subdirectory.

This has been changed in order to avoid accidentally blocking assets
like CSS and JS, for example.

Non existent directories like "/css" and "/web" have been removed,
while missing directories like "/app", "/var" have been added.

eternoendless committed:
Allow jQuery on the Front Office

tomas862 committed:
regenerates assets after rebasing

mickaelandrieu commented:
well... you have hard time with rebasing, maybe @sarjon or @rokas can help a little bit? Not an easy task !

tomas862 commented:
Moved changes to pull request https://github.com/PrestaShop/PrestaShop/pull/9758
